src/Eshopworld.Core/IMessenger.cs

```diff
@@ -5,50 +5,117 @@
     /// Contract that provides a simple way to send and receive messages through callbacks.
     /// </summary>
-    public interface IMessenger : ISendMessages
+    public interface IDoMessages : ISendMessages, ICancelReceive
     {
         /// <summary>
         /// Sets up a call back for receiving any message of type <typeparamref name="T"/>.
-        /// If you try to setup more then one callback to the same message type <typeparamref name="T"/> you'll get an <see cref="InvalidOperationException"/>.
+        /// If you try to setup more then one callback to the same message or event type <typeparamref name="T"/> you'll get an <see cref="InvalidOperationException"/>.
         /// </summary>
         /// <typeparam name="T">The type of the message that we are subscribing to receiving.</typeparam>
         /// <param name="callback">The <see cref="Action{T}"/> delegate that will be called for each message received.</param>
+        /// <param name="batchSize">The size of the batch when reading for a queue - used as the pre-fetch parameter of the message receiver.</param>
         /// <exception cref="InvalidOperationException">Thrown when you attempt to setup multiple callbacks against the same <typeparamref name="T"/> parameter.</exception>
-        void Receive<T>([NotNull] Action<T> callback) where T : IMessage;
+        void Receive<T>(Action<T> callback, int batchSize = 10) where T : class;
+    }
 
+    /// <summary>
+    /// Contract that provides a simple way to send and receive events through callbacks.
+    /// </summary>
+    public interface IDoPubSub : IPublishEvents, ICancelReceive
+    {
         /// <summary>
-        /// Stops receiving a message type by disabling the read pooling on the a message queue.
+        /// Sets up a call back for receiving any event of type <typeparamref name="T"/>.
+        /// If you try to setup more then one callback to the same message or event type <typeparamref name="T"/> you'll get an <see cref="InvalidOperationException"/>.
         /// </summary>
-        /// <typeparam name="T">The type of the message that we are cancelling the receive on.</typeparam>
-        void CancelReceive<T>() where T : IMessage;
+        /// <typeparam name="T">The type of the event that we are subscribing to receiving.</typeparam>
+        /// <param name="callback">The <see cref="Action{T}"/> delegate that will be called for each event received.</param>
+        /// <param name="subscriptionName">The name of the reliable subscription we're doing for this event type.</param>
+        /// <param name="batchSize">The size of the batch when reading for a topic subscription - used as the pre-fetch parameter of the message receiver</param>
+        /// <exception cref="InvalidOperationException">Thrown when you attempt to setup multiple callbacks against the same <typeparamref name="T"/> parameter.</exception>
+        Task Subscribe<T>(Action<T> callback, string subscriptionName, int batchSize = 10) where T : class;
```

Comment:
I understand the constraint but in principle, do we need it? why not allow structs? lots of these "messages" have no need for OOP and have very limited life time and are not passed outside of the "current stack" (somewhere in the code you have a need to rewrap some of the data and push the message out - would be the use case often I think)

Reply:
The constraint is here:
https://github.com/eShopWorld/messaging/blob/master/src/Eshopworld.Messaging/Messenger.cs#L26

If there's a cleaner way at the Observable level to support structs, I'm all in for it, but I don't see any right now! That's why the class constraint bubbles all the way to the top of the messenger API.